File: apple/LookoutMarine/Charts/CoverageMap.swift
//  CoverageMap.swift: where the regions are.
//
//  The lower 48, with Alaska and Hawaii inset. One view cannot hold all three:
//  they span 128 degrees of longitude, and at that scale their latitude span is
//  taller than the sheet. An atlas prints them as insets for the same reason.
//
//  The core projects the coastline and the boxes through the same window, so a
//  panel draws the same on every frame.

import SwiftUI

struct CoverageMap: View {
    let regions: [NoaaRegion]
    let picked: Set<String>
    let enabled: Bool
    /// Each region's real coverage from the catalog. A region with none yet
    /// falls back to its rough extent.
    var coverage: [String: [GeoBox]] = [:]
    let toggle: (String) -> Void

    /// One panel: the ground it covers, and which regions are drawn on it.
    private struct Panel {
        let window: MapWindow
        let panel: Int32
        let label: String?
    }

    private static let main = Panel(
        window: MapWindow(west: -132, east: -64, south: 20, north: 52),
        panel: LOOKOUT_NOAA_PANEL_LOWER48, label: nil)

    private static let insets = [
        Panel(window: MapWindow(west: -172, east: -128, south: 50.5, north: 72),
              panel: LOOKOUT_NOAA_PANEL_ALASKA, label: "Alaska"),
        Panel(window: MapWindow(west: -161, east: -154, south: 18.3, north: 22.6),
              panel: LOOKOUT_NOAA_PANEL_HAWAII, label: "Hawaii"),
    ]

    /// S-52 very shallow water and land, so the picker sits in the app's own
    /// palette.
    private static let water = Chrome.s52("DEPMD").opacity(0.55)
    private static let land = Chrome.s52("LANDA").opacity(0.55)

    var body: some View {
        #if os(macOS)
        // Room for the insets to sit in the Pacific without reaching the
        // coast, at the width the sheet gives the map.
        lower48.overlay(alignment: .bottomLeading) { corners.padding(8) }
        #else
        // Under the map rather than on it. On a screen this narrow the Alaska
        // frame reached the west coast, and a region cannot be picked through
        // the frame of another one.
        VStack(alignment: .leading, spacing: 8) {
            lower48
            corners
        }
        #endif
    }

    private var lower48: some View {
        panel(Self.main)
            .aspectRatio(Self.main.window.aspect, contentMode: .fit)
            .clipShape(RoundedRectangle(cornerRadius: 10, style: .continuous))
            .overlay(RoundedRectangle(cornerRadius: 10, style: .continuous)
                .strokeBorder(Chrome.edge.opacity(0.6), lineWidth: 1))
    }

    private var corners: some View {
        HStack(alignment: .bottom, spacing: 8) {
            inset(Self.insets[0], width: cornerWidth)
            inset(Self.insets[1], width: cornerWidth * 0.54)
        }
    }

    #if os(macOS)
    private var cornerWidth: CGFloat { 134 }
    #else
    /// Smaller than the Mac's. These sit under the map on a phone rather than
    /// in a corner of it, so every point they are tall is a point the regions
    /// below them lose.
    private var cornerWidth: CGFloat { 100 }
    #endif

    /// Alaska and Hawaii keep their own frames, so each reads as itself rather
    /// than as something floating off the coast of Oregon.
    private func inset(_ p: Panel, width: CGFloat) -> some View {
        VStack(alignment: .leading, spacing: 2) {
            if let label = p.label {
                Text(label)
                    .font(.system(size: 9, weight: .medium))
                    .foregroundStyle(Chrome.muted)
            }
            panel(p)
                .frame(width: width, height: width / p.window.aspect)
                .clipShape(RoundedRectangle(cornerRadius: 5, style: .continuous))
                .overlay(RoundedRectangle(cornerRadius: 5, style: .continuous)
                    .strokeBorder(Chrome.edge.opacity(0.7), lineWidth: 1))
        }
    }

    private func panel(_ p: Panel) -> some View {
        GeometryReader { geo in
            ZStack {
                Self.water
                p.window.coast(LOOKOUT_COAST_LAND, in: geo.size).fill(Self.land)
                p.window.coast(LOOKOUT_COAST_LAKE, in: geo.size).fill(Self.water)
                ForEach(regions.filter { $0.panel == p.panel }) { r in
                    box(r, p, in: geo.size)
                }
            }
            .frame(width: geo.size.width, height: geo.size.height)
            .clipped()
        }
        .background(Chrome.panel)
    }

    private func box(_ r: NoaaRegion, _ p: Panel, in size: CGSize) -> some View {
        let on = picked.contains(r.id)
        let shape = RegionShape(boxes: boxes(r), window: p.window)
        // The gesture goes on the shape, so a tap lands on the region's own
        // water rather than on a rectangle around it.
        // Filled, with no stroke. Outlining the path draws every cell box in
        // it, which reads as a mesh over the coast rather than one region.
        return shape
            .fill(Chrome.accent.opacity(on ? 0.5 : 0.16))
            .contentShape(shape)
            .onTapGesture { if enabled { toggle(r.id) } }
    }

    /// The catalog's boxes for a region, or its rough extent until the catalog
    /// is in.
    private func boxes(_ r: NoaaRegion) -> [GeoBox] {
        if let c = coverage[r.id], !c.isEmpty { return c }
        return [GeoBox(west: r.west, south: r.south, east: r.east, north: r.north)]
    }
}


/// A lon/lat window, and the flat rectangle it draws into, in Mercator. The
/// core projects every point.
struct MapWindow {
    let west, east, south, north: Double

    /// Width over height for this window.
    var aspect: CGFloat { CGFloat(lookout_map_aspect(west, east, south, north)) }

    /// Longitude and latitude pairs as points in a rectangle of `size`.
    func points(_ lonlat: [Double], in size: CGSize) -> [CGPoint] {
        let n = lonlat.count / 2
        var xy = [Float](repeating: 0, count: n * 2)
        lookout_map_project(west, east, south, north,
                            Double(size.width), Double(size.height), lonlat, &xy, n)
        return (0..<n).map { CGPoint(x: CGFloat(xy[$0 * 2]), y: CGFloat(xy[$0 * 2 + 1])) }
    }

    /// The coastline rings of one level that reach into this window, as one
    /// path.
    func coast(_ level: Int32, in size: CGSize) -> Path {
        let w = Double(size.width), h = Double(size.height)
        let n = lookout_coastline_rings(level, west, east, south, north, w, h, nil, 0, nil, 0)
        var xy = [Float](repeating: 0, count: n * 2)
        var ends = [UInt32](repeating: 0, count: n / 4)
        lookout_coastline_rings(level, west, east, south, north, w, h, &xy, n, &ends, ends.count)
        func pt(_ i: Int) -> CGPoint { CGPoint(x: CGFloat(xy[i * 2]), y: CGFloat(xy[i * 2 + 1])) }
        var path = Path()
        var start = 0
        for end in ends.map(Int.init) where start < n {
            path.move(to: pt(start))
            for i in start + 1 ..< end { path.addLine(to: pt(i)) }
            path.closeSubpath()
            start = end
        }
        return path
    }
}


/// One region's coverage, as the boxes the catalog states for its coarse
/// cells. Drawn as a single path, so overlapping cells do not stack their fill.
struct RegionShape: Shape {
    let boxes: [GeoBox]
    let window: MapWindow

    func path(in rect: CGRect) -> Path {
        var p = Path()
        let corners = window.points(boxes.flatMap { [$0.west, $0.north, $0.east, $0.south] },
                                    in: rect.size)
        for i in stride(from: 0, to: corners.count - 1, by: 2) {
            let a = corners[i], c = corners[i + 1]
            p.addRect(CGRect(x: min(a.x, c.x), y: min(a.y, c.y),
                             width: max(abs(c.x - a.x), 1.5),
                             height: max(abs(c.y - a.y), 1.5)))
        }
        return p
    }
}


/// The regions as pills under the map: the name, and the mark for a chosen one.
struct NoaaRegionPills: View {
    let regions: [NoaaRegion]
    let picked: Set<String>
    let enabled: Bool
    /// How much of each region is on the device. A pill for water already held
    /// says so, because the tick then means "keep this" rather than "fetch it".
    var state: [String: NoaaRegionState] = [:]
    let toggle: (String) -> Void

    var body: some View {
        FlowRow(spacing: 8, lineSpacing: 8) {
            ForEach(regions) { r in
                pill(r)
            }
        }
    }

    private func pill(_ r: NoaaRegion) -> some View {
        let on = picked.contains(r.id)
        return Button { toggle(r.id) } label: {
            HStack(spacing: 6) {
                if on {
                    Image(systemName: "checkmark")
                        .font(.system(size: 10, weight: .bold))
                }
                Text(r.name)
                    .font(.system(size: 12.5, weight: on ? .semibold : .regular))
                if let note = badge(r) {
                    Text(note)
                        .font(.system(size: 11))
                        .monospacedDigit()
                        .opacity(0.75)
                }
            }
            .foregroundStyle(on ? Color.white : Chrome.ink)
            .padding(.horizontal, 13)
            .frame(height: pillHeight)
            .background(on ? Chrome.accent : Chrome.surface, in: Capsule())
            .overlay(Capsule().strokeBorder(on ? .clear : Chrome.edge.opacity(0.8), lineWidth: 1))
        }
        .buttonStyle(.plain)
        .disabled(!enabled)
        .opacity(enabled ? 1 : 0.5)
        .help(r.blurb)
        .accessibilityLabel(label(r))
        .accessibilityAddTraits(on ? [.isButton, .isSelected] : .isButton)
        .accessibilityIdentifier("region-\(r.id)")
    }

    /// What of this region is here, in the pill. Nothing before the catalog is
    /// read, and nothing for water with none of it on the device.
    /// Only the whole of a region is stated. NOAA files cells across district
    /// lines, so downloading one region installs some of its neighbour's, and
    /// a count of those read as a transfer that had stopped part way.
    private func badge(_ r: NoaaRegion) -> String? {
        guard let st = state[r.id], st.allHeld else { return nil }
        return "installed"
    }

    private func label(_ r: NoaaRegion) -> String {
        guard let st = state[r.id] else { return "\(r.name). \(r.blurb)" }
        if st.allHeld { return "\(r.name). \(r.blurb). All \(st.held) charts installed." }
        return "\(r.name). \(r.blurb). \(st.cells) charts to download."
    }

    #if os(macOS)
    private var pillHeight: CGFloat { 30 }
    #else
    // A thumb needs the full target.
    private var pillHeight: CGFloat { 44 }
    #endif
}


/// The regions as rows, for a narrow screen.
///
/// Nine pills wrap to three ragged lines on a phone and leave the rest of the
/// page empty under them. A row is the width of the screen, says which water
/// the region covers rather than making the map answer that, and is a target a
/// thumb cannot miss.
struct NoaaRegionRows: View {
    let regions: [NoaaRegion]
    let picked: Set<String>
    let enabled: Bool
    let toggle: (String) -> Void

    var body: some View {
        VStack(spacing: 0) {
            ForEach(Array(regions.enumerated()), id: \.element.id) { i, r in
                if i > 0 { Divider().padding(.leading, 14) }
                row(r)
            }
        }
        .background(Chrome.surface,
                    in: RoundedRectangle(cornerRadius: 12, style: .continuous))
        .overlay(RoundedRectangle(cornerRadius: 12, style: .continuous)
            .strokeBorder(Chrome.edge.opacity(0.8), lineWidth: 1))
        .opacity(enabled ? 1 : 0.5)
    }

    private func row(_ r: NoaaRegion) -> some View {
        let on = picked.contains(r.id)
        return Button { toggle(r.id) } label: {
            HStack(spacing: 12) {
                VStack(alignment: .leading, spacing: 2) {
                    Text(r.name)
                        .font(.system(size: 15, weight: on ? .semibold : .regular))
                        .foregroundStyle(Chrome.ink)
                    Text(r.blurb)
                        .font(.system(size: 12.5))
                        .foregroundStyle(Chrome.muted)
                        .multilineTextAlignment(.leading)
                        .fixedSize(horizontal: false, vertical: true)
                }
                Spacer(minLength: 8)
                Image(systemName: on ? "checkmark.circle.fill" : "circle")
                    .font(.system(size: 20))
                    .foregroundStyle(on ? Chrome.accent : Chrome.ink.opacity(0.25))
            }
            .padding(.horizontal, 14)
            .padding(.vertical, 11)
            .frame(minHeight: 56)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .disabled(!enabled)
        .accessibilityLabel("\(r.name). \(r.blurb)")
        .accessibilityAddTraits(on ? [.isButton, .isSelected] : .isButton)
        .accessibilityIdentifier("region-\(r.id)")
    }
}

File: android/app/src/main/java/org/beetlebug/lookout/charts/CoverageMap.kt
package org.beetlebug.lookout.charts

import org.beetlebug.lookout.Lookout
import org.beetlebug.lookout.hud.Chrome

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.toSize
import kotlin.math.max
import kotlin.math.min

/**
 * Where the regions are.
 *
 * The lower 48, with Alaska and Hawaii under it. One view cannot hold all
 * three: they span 128 degrees of longitude, and at that scale their latitude
 * span is taller than the screen. An atlas prints them separately for the same
 * reason.
 *
 * The map is a picture of the pick, not the control for it. Tapping a region
 * on it works, and the rows under it are how a thumb does the same thing.
 */

/** The ground one panel covers. The core projects onto it, in Mercator. */
class MapWindow(val west: Double, val south: Double, val east: Double, val north: Double) {
    /** Width over height, so a panel keeps its shape at any size. */
    val aspect: Float get() = Lookout.mapAspect(west, east, south, north).toFloat()

    /** Longitude and latitude pairs as x and y pairs in a rectangle of [size]. */
    fun project(lonlat: DoubleArray, size: Size): FloatArray =
        Lookout.mapProject(west, east, south, north, size.width, size.height, lonlat)

    /** Each box's corners, north-west then south-east, projected. */
    fun corners(boxes: List<NoaaController.Box>, size: Size): FloatArray {
        val lonlat = DoubleArray(boxes.size * 4)
        boxes.forEachIndexed { i, b ->
            lonlat[i * 4] = b.west
            lonlat[i * 4 + 1] = b.north
            lonlat[i * 4 + 2] = b.east
            lonlat[i * 4 + 3] = b.south
        }
        return project(lonlat, size)
    }
}

private val MAIN = MapWindow(-132.0, 20.0, -64.0, 52.0)
private val ALASKA = MapWindow(-172.0, 50.5, -128.0, 72.0)
private val HAWAII = MapWindow(-161.0, 18.3, -154.0, 22.6)

/** S-52 very shallow water and land, so the picker sits in the chart's own
 *  palette rather than the system's. */
@Composable
private fun water() = Chrome.s52("DEPMD").copy(alpha = 0.55f)

@Composable
private fun land() = Chrome.s52("LANDA").copy(alpha = 0.55f)

@Composable
fun CoverageMap(
    regions: List<NoaaController.Region>,
    picked: Set<String>,
    enabled: Boolean,
    coverage: Map<String, List<NoaaController.Box>>,
    onToggle: (String) -> Unit,
    modifier: Modifier = Modifier,
) {
    val accent = MaterialTheme.colorScheme.primary
    val edge = MaterialTheme.colorScheme.outlineVariant

    Column(modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(8.dp)) {
        panel(MAIN, Lookout.PANEL_LOWER48, regions, picked, enabled, coverage, accent, edge, onToggle,
              Modifier.fillMaxWidth().aspectRatio(MAIN.aspect))
        // Alaska and Hawaii keep their own frames, under the map rather than
        // in a corner of it: on a narrow screen a frame in the corner covers
        // the west coast, and a region cannot be picked through another one.
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            inset("Alaska", ALASKA, Lookout.PANEL_ALASKA, regions, picked, enabled, coverage,
                  accent, edge, onToggle, 104.dp)
            inset("Hawaii", HAWAII, Lookout.PANEL_HAWAII, regions, picked, enabled, coverage,
                  accent, edge, onToggle, 56.dp)
        }
    }
}

@Composable
private fun inset(
    label: String,
    window: MapWindow,
    which: Int,
    regions: List<NoaaController.Region>,
    picked: Set<String>,
    enabled: Boolean,
    coverage: Map<String, List<NoaaController.Box>>,
    accent: Color,
    edge: Color,
    onToggle: (String) -> Unit,
    width: androidx.compose.ui.unit.Dp,
) {
    Column {
        Text(
            label,
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Spacer(Modifier.height(2.dp))
        panel(window, which, regions, picked, enabled, coverage, accent, edge, onToggle,
              Modifier.width(width).height(width / window.aspect))
    }
}

@Composable
private fun panel(
    window: MapWindow,
    which: Int,
    regions: List<NoaaController.Region>,
    picked: Set<String>,
    enabled: Boolean,
    coverage: Map<String, List<NoaaController.Box>>,
    accent: Color,
    edge: Color,
    onToggle: (String) -> Unit,
    modifier: Modifier,
) {
    val mine = remember(regions, which) { regions.filter { it.panel == which } }
    val water = water()
    val land = land()
    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(10.dp),
        color = water,
        border = BorderStroke(1.dp, edge),
    ) {
        Canvas(
            Modifier.fillMaxSize().clip(RoundedCornerShape(10.dp)).pointerInput(enabled, mine, coverage) {
                if (!enabled) return@pointerInput
                detectTapGestures { at ->
                    val hit = mine.firstOrNull { r ->
                        val c = window.corners(boxesOf(r, coverage), size.toSize())
                        (0 until c.size / 4).any { i ->
                            val x0 = c[i * 4]
                            val y0 = c[i * 4 + 1]
                            val x1 = c[i * 4 + 2]
                            val y1 = c[i * 4 + 3]
                            at.x in min(x0, x1)..max(x0, x1) && at.y in min(y0, y1)..max(y0, y1)
                        }
                    }
                    if (hit != null) onToggle(hit.id)
                }
            },
        ) {
            drawLand(window, land, water)
            for (r in mine) {
                val on = picked.contains(r.id)
                drawRegion(boxesOf(r, coverage), window, accent.copy(alpha = if (on) 0.5f else 0.16f))
            }
        }
    }
}

/** The catalog's boxes for a region, or its rough extent until the catalog is
 *  in. One rectangle claims water a district does not cover, so the boxes are
 *  used the moment they exist. */
private fun boxesOf(
    r: NoaaController.Region,
    coverage: Map<String, List<NoaaController.Box>>,
): List<NoaaController.Box> {
    val c = coverage[r.id]
    if (!c.isNullOrEmpty()) return c
    return listOf(NoaaController.Box(r.west, r.south, r.east, r.north))
}

/**
 * The coastline rings that reach into this window. Land and lakes are drawn in
 * level order, so a lake paints water back over the land it sits in.
 */
private fun DrawScope.drawLand(window: MapWindow, land: Color, water: Color) {
    for (level in intArrayOf(Lookout.COAST_LAND, Lookout.COAST_LAKE)) {
        val rings = Lookout.coastlineRings(level, window.west, window.east, window.south,
                                           window.north, size.width, size.height)
        val xy = rings[0] as FloatArray
        val ends = rings[1] as IntArray
        val path = Path()
        var start = 0
        for (end in ends) {
            path.moveTo(xy[start * 2], xy[start * 2 + 1])
            for (i in start + 1 until end) path.lineTo(xy[i * 2], xy[i * 2 + 1])
            path.close()
            start = end
        }
        drawPath(path, if (level == Lookout.COAST_LAND) land else water)
    }
}

/** One region's coverage, as a single path so overlapping cells do not stack
 *  their fill into a darker patch. */
private fun DrawScope.drawRegion(
    boxes: List<NoaaController.Box>,
    window: MapWindow,
    color: Color,
) {
    val path = Path()
    val c = window.corners(boxes, size)
    for (i in 0 until c.size / 4) {
        val x0 = c[i * 4]
        val y0 = c[i * 4 + 1]
        val x1 = c[i * 4 + 2]
        val y1 = c[i * 4 + 3]
        path.addRect(
            androidx.compose.ui.geometry.Rect(
                Offset(min(x0, x1), min(y0, y1)),
                Size(max(max(x1 - x0, x0 - x1), 1.5f), max(max(y1 - y0, y0 - y1), 1.5f)),
            ),
        )
    }
    drawPath(path, color)
}
